### PR TITLE
Feat (Shutter): Add airdrop and delegation decoder

### LINKED ISSUE #7357: Add Shutter token, airdrop claim decoding and airdrop detection
## Abstract

ShutterDAO [launched](https://blog.shutter.network/a-proposed-blueprint-for-launching-a-shutter-dao/) and we can support it better in rotki and help rotki users find better visualize shutter related transactions and find out about the airdrop and how they can claim it.

## Task

- Add the token to the global DB: https://etherscan.io/token/0xe485e2f1bab389c08721b291f6b59780fec83fd7. Though we have no oracle, price, or icon in coingecko so won't make much of a difference now.
- Decode airdrop claiming transaction and delegation transactions. Can find all of them in the token's page so pick any example to decode.
    * Example claiming transaction: https://etherscan.io/tx/0x497d43b4a84c73ff570fc3034f0da54c64e99043c501030138f9f555715d9769
    * Example delegating transaction: https://etherscan.io/tx/0x4102909033e878dd68c2f33e54c014700772d32f8993a717532a14c95f904056
- Add the airdrop to our list of airdrops and detect it for any users whose tracked addresses can claim it.
    * The list of users is in these CSVs here: https://github.com/shutter-network/shutter-dao-claiming-app-data/tree/main/vestings/assets/1. We probably only need few of these columns so we could just combine them and upload them to our data repo
    * There is a cutoff date, so that should be taken into account. It's Sun Jul 14 2024 23:33:20 UTC. Could also be combined with the calendar feature to warn users if they have not yet claimed and they are eligile: https://github.com/rotki/rotki/issues/7074

### PR DESCRIPTION
**Pull request by OjusWiZard**
Closes #7357

## Checklist

- [x] Add SHU token in globaldb (https://github.com/rotki/assets/pull/127)
- [x] Add Shutter decoder for decoding SHU airdrops and changing delegates
- [x] Add SHU airdrop detection (https://github.com/rotki/data/pull/45)
- [x] Update changelog and add tests

SQL Diff of `rotkehlchen/data/global.db`:
```sql
INSERT INTO assets(rowid,identifier,name,type) VALUES(5435,'eip155:1/erc20:0xe485E2f1bab389C08721B291f6b59780feC83Fd7','SHU','C');
INSERT INTO common_asset_details(rowid,identifier,symbol,coingecko,cryptocompare,forked,started,swapped_for) VALUES(5435,'eip155:1/erc20:0xe485E2f1bab389C08721B291f6b59780feC83Fd7','SHU','','',NULL,1705431563,NULL);
INSERT INTO evm_tokens(rowid,identifier,token_kind,chain,address,decimals,protocol) VALUES(4534,'eip155:1/erc20:0xe485E2f1bab389C08721B291f6b59780feC83Fd7','A',1,'0xe485E2f1bab389C08721B291f6b59780feC83Fd7',18,NULL);
```

### CODE REVIEW

File: rotkehlchen/chain/ethereum/airdrops.py [hunk 1]
```diff
@@ -290,10 +317,10 @@ def get_airdrop_data(name: str, data_dir: Path) -> Iterator[list[str]]:
         # if not cached, get it from the gist
         try:
             response = requests.get(url=AIRDROPS[name][0], timeout=(30, 100))  # a large read timeout is necessary because the queried data is quite large  # noqa: E501
+            content = response.text
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/airdrops.py:320]: why did you move this line?

File: rotkehlchen/chain/ethereum/airdrops.py [hunk 2]
```diff
@@ -382,6 +409,8 @@ def check_airdrops(
     data_dir = database.user_data_dir.parent
     airdrop_tuples = []
     for protocol_name, airdrop_data in AIRDROPS.items():
+        if airdrop_data.cutoff_time is not None and time() > airdrop_data.cutoff_time:
+            continue
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/airdrops.py:413]: don't execute the function on every loop iteration. Python has to move the context and is expensive. Create a variable outside and assign it to time(). Also shouldn't you use `ts_now()`

Comment (yabirgb) [reply rotkehlchen/chain/ethereum/airdrops.py:413]: Can you also add a log debug here saying that you skip the airdrop?

File: rotkehlchen/chain/ethereum/modules/shutter/constants.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from typing import Final
+from rotkehlchen.chain.evm.types import string_to_evm_address
+
+
+SHU_CLAIMED = b'|\xd0\x91\xe2\xff?\x9f\xb5\xd2\xaf[\x8f\xf0\xf5\x94\xb6\x96=\xb4N\x14?\xd4q\xa12\xb6IS_\x126'  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/shutter/constants.py:5]: make it final

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+        amount = vesting_pool_address = None
+        for tx_log in context.all_logs:
+            if (
+                tx_log.address == self.shu.evm_address and
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                hex_or_bytes_to_address(tx_log.topics[1]) == SHUTTER_AIDROP_CONTRACT
+            ):
+                amount = token_normalized_value(token_amount=hex_or_bytes_to_int(tx_log.data), token=self.shu)  # noqa: E501
+                vesting_pool_address = hex_or_bytes_to_address(tx_log.topics[2])
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:63]: isn't this address the same for everyone?

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+        return DecodingOutput(event=self.base.make_event_from_transaction(
+            transaction=context.transaction,
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=self.shu,
+            balance=Balance(amount=FVal(amount)),
+            location_label=hex_or_bytes_to_address(context.tx_log.topics[2]),
+            notes=f'Claim {amount} SHU from shutter airdrop and deposit it to {vesting_pool_address}',  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:81]: I believe the term is 

```suggestion
            notes=f'Claim  {amount} SHU vested from shutter airdrop and deposit it to {vesting_pool_address}',  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            counterparty=CPT_SHUTTER,
+            address=context.transaction.to_address,
+        ))
+
+    def _decode_delegation_change(self, context: DecoderContext) -> DecodingOutput:
+        """Decode a change in the delegated address"""
+        delegator = hex_or_bytes_to_address(context.tx_log.topics[1])
+        from_delegate = hex_or_bytes_to_address(context.tx_log.topics[2])
+        to_delegate = hex_or_bytes_to_address(context.tx_log.topics[3])
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:90]: isn't delegator always one of the other two? I'm thinking that maybe you can remove one of the if checks

File: rotkehlchen/tests/unit/decoders/test_shutter.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            counterparty=CPT_GAS,
+        ), EvmEvent(
+            tx_hash=tx_hash,
+            sequence_index=132,
+            timestamp=timestamp,
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=A_SHU,
+            balance=Balance(amount=FVal('1086.95652173913')),
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_shutter.py:50]: move this amount also to a variable like you did for the gas

File: rotkehlchen/tests/unit/decoders/test_shutter.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            tx_hash=tx_hex,
+            sequence_index=479,
+            timestamp=timestamp,
+            location=Location.ETHEREUM,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.GOVERNANCE,
+            asset=A_SHU,
+            balance=Balance(),
+            location_label=ethereum_accounts[0],
+            notes=f'Change SHU Delegate of 0x561075538e7B20613Aa0C3fAa8459ac293011584 from {ethereum_accounts[0]} to 0x9Cc9C7F874eD77df06dCd41D95a2C858cd2a2506',  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_shutter.py:95]: I believe you can remove this `of 0x.. ` part... this one is the vesting contract if I'm not mistaken

File: docs/changelog.rst [hunk 9]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :feature:`7357` rotki will now properly detect and decode Shutter token airdrops and delegations.
```
Comment (yabirgb) [inline docs/changelog.rst:5]:
```suggestion
* :feature:`7357` rotki will now properly detect and decode Shutter token airdrop and delegation.
```

**Review by yabirgb — COMMENTED** [review]
some comments

File: rotkehlchen/chain/ethereum/airdrops.py (hunk 1 above)
Comment (OjusWiZard) [reply rotkehlchen/chain/ethereum/airdrops.py:320]: the `content` variable was possibly unbound in except block if we kept it in its try block
image

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py (hunk 6 above)
Comment (OjusWiZard) [reply rotkehlchen/chain/ethereum/modules/shutter/decoder.py:90]: I can't find any transaction where it's the same. It's the vesting pool address in all the transactions that I've checked.

File: rotkehlchen/tests/unit/test_ethereum_airdrops.py [hunk 10]
```diff
@@ -131,6 +141,26 @@ def mock_requests_get(url: str, timeout: int):  # pylint: disable=unused-argumen
...
+    freezer.move_to(datetime.datetime.fromtimestamp(  # after cutoff time of shutter
+        AIRDROPS['shutter'].cutoff_time + 1, tz=datetime.UTC),
+    )
+    with (
+        patch('rotkehlchen.chain.ethereum.airdrops.SMALLEST_AIRDROP_SIZE', 1),
+        patch('rotkehlchen.chain.ethereum.airdrops.requests.get', side_effect=mock_requests_get),
+    ):
+        data = check_airdrops(
+            addresses=[TEST_ADDR2], database=database,
+            tolerance_for_amount_check=tolerance_for_amount_check,
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_ethereum_airdrops.py:160]:
```suggestion
            addresses=[TEST_ADDR2],
            database=database,
            tolerance_for_amount_check=tolerance_for_amount_check,
```

File: rotkehlchen/tests/unit/decoders/test_shutter.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    tx_hex = deserialize_evm_tx_hash('0x3f20929de51baefdc688997a05bde1e32120cb7d4e0fda5da3963b1a620d0a8b')  # noqa: E501
+    tx_hash = deserialize_evm_tx_hash(tx_hex)
+    events, _ = get_decoded_events_of_transaction(
+        evm_inquirer=ethereum_inquirer,
+        database=database,
+        tx_hash=tx_hex,
+    )
+    timestamp = TimestampMS(1706703155000)
+    gas_amount_str = '0.006946508605618104'
+    claimed_amount = '1086.95652173913'
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_shutter.py:29]:
```suggestion
    gas_amount_str, claimed_amount = '0.006946508605618104', '1086.95652173913'
```

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+        return DecodingOutput(
+            event=self.base.make_event_from_transaction(
+                transaction=context.transaction,
+                tx_log=context.tx_log,
+                event_type=HistoryEventType.INFORMATIONAL,
+                event_subtype=HistoryEventSubType.GOVERNANCE,
+                asset=self.shu,
+                balance=Balance(),
+                location_label=event_address,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:106]: What are you doing man? I don't think you understand what location label is. 

This is the address that the event touches.It's the address I filter by when I filter my events by address.

So if I delegate to someone then this is my address. The way you made it is ... either the one I used to delegate to, or the one I will delegate to. Which is sort of stupid .... no?

User `user_address` here which should be `transaction.from_address`. Since it's used nowhere else no need to make an extra variable for it I guess.

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            balance=Balance(amount=FVal(amount)),
+            location_label=hex_or_bytes_to_address(context.tx_log.topics[2]),
+            notes=f'Claim {amount} SHU vested from shutter airdrop and deposit it to {vesting_pool_address}',  # noqa: E501
+            counterparty=CPT_SHUTTER,
+            address=context.transaction.to_address,
+        ))
+
+    def _decode_delegation_change(self, context: DecoderContext) -> DecodingOutput:
+        """Decode a change in the delegated address"""
+        from_delegate = hex_or_bytes_to_address(context.tx_log.topics[2])
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:88]: The delegator here is the one who delegates. But since the delegator is the vesting contract in all these cases and we can't know who owns the vesting contract we are forced to make an assumption. `context.transaction.from_address` being the user address.

And this assumption is probaly always gonna hold since you are matching this with input data. So the caller, is the ones who owns the vesting contract.

File: rotkehlchen/tests/unit/decoders/test_shutter.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            notes=f'Change SHU Delegate from {ethereum_accounts[0]} to 0x9Cc9C7F874eD77df06dCd41D95a2C858cd2a2506',  # noqa: E501
+            counterparty=CPT_SHUTTER,
+        ),
+    ]
+    assert events == expected_events
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [['0x36bE40f9bd613dFf47294E7e10D4cae072E06A2D']])
+def test_shu_delegation_to_self(database, ethereum_inquirer, ethereum_accounts):
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_shutter.py:104]: Test delegation to someone else. Not to self. You have completely misunderstood what delegation is.

File: rotkehlchen/tests/unit/decoders/test_shutter.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+            counterparty=CPT_SHUTTER,
+            address=SHUTTER_AIDROP_CONTRACT,
+        ),
+    ]
+    assert events == expected_events
+
+
+@pytest.mark.vcr()
+@pytest.mark.parametrize('ethereum_accounts', [['0x36bE40f9bd613dFf47294E7e10D4cae072E06A2D']])
+def test_shu_delegation_from_self(database, ethereum_inquirer, ethereum_accounts):
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/decoders/test_shutter.py:62]: Remove one of these delegation tests. They are useless. You did not understand what delegation is.

**Review by LefterisJP — COMMENTED** [review]
Really disappointing.

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+            protocol=CPT_SHUTTER,
+            proposals_url='https://shutterdao.discourse.group/c/shutter-dao/dao-proposals/15',
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:47]: This is also wrong. There is no on-chain proposals in shutter at the moment. So make it an optional argument and don't populate it here.

If it was an onchain proposal, this would not be how the link looks like. Again as I said ask what stuff are about before making PRs.

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+        return DecodingOutput(event=self.base.make_event_from_transaction(
+            transaction=context.transaction,
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=self.shu,
+            balance=Balance(amount=FVal(amount)),
+            location_label=hex_or_bytes_to_address(context.tx_log.topics[2]),
+            notes=f'Claim {amount} SHU vested from shutter airdrop and deposit it to {vesting_pool_address}',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:81]:
```suggestion
            notes=f'Claim {amount} SHU from shutter airdrop into the vesting contract: {vesting_pool_address}',  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+        amount = vesting_pool_address = None
+        for tx_log in context.all_logs:
+            if (
+                tx_log.address == self.shu.evm_address and
+                tx_log.topics[0] == ERC20_OR_ERC721_TRANSFER and
+                hex_or_bytes_to_address(tx_log.topics[1]) == SHUTTER_AIDROP_CONTRACT
+            ):
+                amount = token_normalized_value(token_amount=hex_or_bytes_to_int(tx_log.data), token=self.shu)  # noqa: E501
+                vesting_pool_address = hex_or_bytes_to_address(tx_log.topics[2])
```
[the remark this answers is not in the document]
Comment (OjusWiZard) [reply rotkehlchen/chain/ethereum/modules/shutter/decoder.py:63]: It's the [contract name](https://etherscan.io/address/0x22714964e6b9F17798A9e1AD3f2BAb87279876FC#code#F25#L11), but changing nw.

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.shu = A_SHU.resolve_to_evm_token()
+
+    def _decode_shutter_claim(self, context: DecoderContext) -> DecodingOutput:
+        if not (
+            context.tx_log.topics[0] == REDEEMED_VESTING and
+            self.base.is_tracked(hex_or_bytes_to_address(context.tx_log.topics[2]))
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:52]: I told you in the old review: https://github.com/rotki/rotki/pull/7388/files#r1475181550

**keep the user address**

```suggestion
            self.base.is_tracked(user_address := hex_or_bytes_to_address(context.tx_log.topics[2])))
```

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+            return DEFAULT_DECODING_OUTPUT
+
+        return DecodingOutput(event=self.base.make_event_from_transaction(
+            transaction=context.transaction,
+            tx_log=context.tx_log,
+            event_type=HistoryEventType.INFORMATIONAL,
+            event_subtype=HistoryEventSubType.NONE,
+            asset=self.shu,
+            balance=Balance(amount=FVal(amount)),
+            location_label=hex_or_bytes_to_address(context.tx_log.topics[2]),
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:84]: Why did you mark this as resolved. You are not using the user address from the variable as I said, but you are recalculating it.
```suggestion
            location_label=user_address,
```

File: rotkehlchen/chain/ethereum/modules/shutter/decoder.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        ))
+
+    # -- DecoderInterface methods
+
+    def addresses_to_decoders(self) -> dict[ChecksumEvmAddress, tuple[Any, ...]]:
+        return {
+            SHUTTER_AIDROP_CONTRACT: (self._decode_shutter_claim,),
+        }
+
+    def decoding_by_input_data(self) -> dict[bytes, dict[bytes, Callable]]:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/shutter/decoder.py:116]: Why are you using decoding by input data? Decoding by input_data should be avoided as it's much wider-catching than the other rules, and as such also makes the entire decoding process slower.

WE only use it when we have no idea the contract that will generate the logs. As we do with SAFE. Since anyone can deploy a SAFE contract.

Here youdon't need this. You just need to use the `addresses_to_decoders` with `0xe485E2f1bab389C08721B291f6b59780feC83Fd7`  (shutter token address), with the `DelegateChanged` log.

**Review by LefterisJP — APPROVED** [review]
LGTM. Failed tests are due to 503 from blockstream
